ports-of-call/portable_arrays.hpp

```diff
@@ -26,73 +26,145 @@
 
 #include "portability.hpp"
 #include <algorithm>
+#include <array>
 #include <assert.h>
 #include <cstddef> // size_t
 #include <cstring> // memset()
+#include <functional>
+#include <numeric>
+#include <type_traits>
 #include <utility> // swap()
 
+// maximum number of dimensions
+constexpr std::size_t MAXDIM = 6;
+// array type of dimensions/strides
+using narr = std::array<std::size_t, MAXDIM>;
+
+namespace detail {
+
+// set_value end-case
+template <std::size_t Ind, typename NX>
+PORTABLE_INLINE_FUNCTION void set_value(narr &ndat, NX value) {
+  ndat[Ind] = value;
+}
+
+// set_valueS end case, dispatch to set_value
+template <std::size_t Ind, typename NX>
+PORTABLE_INLINE_FUNCTION void set_values(narr &ndat, NX value) {
+  set_value<Ind>(ndat, value);
+}
+
+// set_values general case, set `Ind` position of array (tracking from RIGHT)
+// with head of size list `NX`
+template <std::size_t Ind, typename NX, typename... NXs>
+PORTABLE_INLINE_FUNCTION void set_values(narr &ndat, NX value, NXs... nxs) {
+  set_value<Ind>(ndat, value);
+  set_values<Ind - 1>(ndat, nxs...);
+}
+
+// compute_index base case, i.e. fastest moving index
+template <std::size_t Ind>
+PORTABLE_INLINE_FUNCTION size_t compute_index(const narr &nd,
+                                              const size_t index) {
+  return index;
+}
+
+// compute_index general case, computing slower moving index strides
+template <std::size_t Ind, typename... Tail>
+PORTABLE_INLINE_FUNCTION size_t compute_index(const narr &nd,
+                                              const size_t index,
+                                              const Tail... tail) {
+  return index * nd[Ind] + compute_index<Ind + 1>(nd, tail...);
+}
```

Comment:
This is nice---assuming the array on device issues work.

Reply:
Tests complete on `volta-x86`, though we need to put together some more tests for `ports-of-call`